http: add CDN-Loop header value parser and utilities

justin-mp opened the pull request:
The CDN-Loop header is defined by RFC 8586 and is a comma-separated
list of values.  However, it’s not possible to separate the values by
splitting on `','` because the values may have parameters, which
themselves may contain strings, which may contain commas.  Thus, we
need a real parser to extract the values.

This pull request contains a CDN-Loop header parser and a utility that
uses the parser to count how many times a particular CDN’s identifier
has been seen in the header.

The parser and the utility are the first two steps towards having a
CDN-Loop aware filter.  See #8183.

Risk: Low
Testing: bazel test //test/extensions/filters/http/cdn:cdn_loop_parser_test //test/extensions/filters/http/cdn:cdn_loop_utils_test
Documentation: N/A
Release Notes: N/A

Signed-off-by: Justin Mazzola Paluska <justinmp@google.com>

justin-mp commented:
@alyssawilk : I added you as a code owner in addition to @penguingao and me.  Please let me know if there's a better maintainer to add as a code owner.

alyssawilk commented on source/extensions/filters/http/cdn/cdn_loop_utils.cc:
```diff
@@ -0,0 +1,34 @@
+namespace HttpFilters {
+namespace Cdn {
+
+StatusOr<int> countCdnLoopOccurrences(absl::string_view header, absl::string_view cdn_id) {
```
Glancing at the spec, I think we'd want to handle HTTP/1.1. style multi-entry headers, where you can have
Foo: bar
Foo: eep
not just 
Foo: bar,eep
Are you assuming any such headers would be appended together or should this take 
vector<absl::string_view> header_values ?

justin-mp replied:
I'm assuming that any such headers would be appended together.  I'm targeting what I get out of `Envoy::Http::HeaderMap`, which returns a `HeaderEntry`, which returns a single `HeaderString&` as its value.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+// RFC 5234 Appendix B.1 says:
+//
+// ALPHA          =  %x41-5A / %x61-7A   ; A-Z / a-z
+constexpr bool IsAlpha(char c) {
```
No, because the functions are locale-aware.  For example, if one runs Envoy in anything other than the `"C"` locale, the `isalpha` will have a different set of allowed "alpha" characters than the RFC.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.h:
```diff
@@ -0,0 +1,293 @@
+// and the parser defined here are:
+//
+// 1. the parser has a more lax interpretation of what's a valid uri-host. See
+//     ParseCdnId for details.
```
`cdn; cdn="cdn,cdn"` is legit according to the RFC.  The example in the RFC makes use of quoted strings, though not the evil `"cdn,cdn"` one.  My first guess was to just split on `','` and `';'`, but quoted strings makes that impossible.

penguingao commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+  if (StatusOr<ParseContext> ipv6 = ParsePlausibleIpV6(context); ipv6) {
+    context.update(*ipv6);
+  } else {
+    if (StatusOr<ParseContext> token = ParseToken(context); !token) {
```
it looks to me that uri-host contains more characters than token, so we may be unable to parse an otherwise valid cdn-id.

More specifically: reg-name    = *( unreserved / pct-encoded / sub-delims )

pct-encoded is not parsable. sub-delims contains more characters than token too.

justin-mp replied:
Valid pct-encoded is parsable by token.  RFC 3986 Appendix A says:
```
pct-encoded   = "%" HEXDIG HEXDIG
```
while token allows "%" and HEXDIG.  I'll add a test to make sure this sticks.

sub-delims is tricky because allowing all sub-delims makes the grammar ambiguous.  RFC 3986 Appendix A says:
```
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```
In particular, `","` and `";"` make the grammar ambiguous since we don't know if the comma is part of the `uri-host` or if it delimits the start of the parameters (`";"`) or the start of the next entry.  You're right that token doesn't cover some of the other entries.  Let me dig into this a bit more to see if it makes sense to add a special parser to handle them.

justin-mp replied:
I read through the archives of the ietf-http-wg@w3.org mailing list where RFC 8586 was discussed (https://lists.w3.org/Archives/Public/ietf-http-wg/2018OctDec/thread.html and https://lists.w3.org/Archives/Public/ietf-http-wg/2019JanMar/thread.html for the curious, search for "cdn") to dig into the details a bit more.  Initially, the authors only wanted to allow tokens but were persuaded to allow hosts.

RFC 8586 says:
```
The cdn-id identifies the CDN using either a hostname under its control or a pseudonym.
```
The phrase "hostname under its control" corresponds to a `reg-name` in [RFC 3986 Section 3.2.2](https://tools.ietf.org/html/rfc3986#section-3.2.2]).  RFC 3986 says:
```
   A host identified by a registered name is a sequence of characters
   usually intended for lookup within a locally defined host or service
   name registry, though the URI's scheme-specific semantics may require
   that a specific registry (or fixed name table) be used instead.  The
   most common name registry mechanism is the Domain Name System (DNS).
   A registered name intended for lookup in the DNS uses the syntax
   defined in Section 3.5 of [RFC1034] and Section 2.1 of [RFC1123].
   Such a name consists of a sequence of domain labels separated by ".",
   each domain label starting and ending with an alphanumeric character
   and possibly also containing "-" characters.  The rightmost domain
   label of a fully qualified domain name in DNS may be followed by a
   single "." and should be if it is necessary to distinguish between
   the complete domain name and some local domain.

...

   This specification does not mandate a particular registered name
   lookup technology and therefore does not restrict the syntax of reg-
   name beyond what is necessary for interoperability.  Instead, it
   delegates the issue of registered name syntax conformance to the
   operating system of each application performing URI resolution, and
   that operating system decides what it will allow for the purpose of
   host identification.  A URI resolution implementation might use DNS,
   host tables, yellow pages, NetInfo, WINS, or any other system for
   lookup of registered names.  However, a globally scoped naming
   system, such as DNS fully qualified domain names, is necessary for
   URIs intended to have global scope.  URI producers should use names
   that conform to the DNS syntax, even when use of DNS is not
   immediately apparent, and should limit these names to no more than
   255 characters in length.
```

Given that Envoy assumes it's on the current Internet, we should assume DNS names.

The current implementation admits DNS names (and more), so I think it's appropriate here.

Do you want me to update the comment for `parseCdnId` with this discussion?

justin-mp replied:
Done.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+ParseContext ParseOptionalWhitespace(const ParseContext& input) {
+  ParseContext context = input;
+  while (!context.atEnd()) {
+    char c = context.peek();
```
Done.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+
+// RFC 7230 Section 3.2.6 defines qdtext as:
+//
+// qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
```
Done

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+      continue;
+    }
+
+    // sub-expressions have parsed as much as they can
```
The break is vestigial.  I've removed it since, as you point out, it's unreachable.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+    char c = context.peek();
+    if (c == '!' || c == '#' || c == '$' || c == '%' || c == '&' || c == '\'' || c == '*' ||
+        c == '+' || c == '-' || c == '.' || c == '^' || c == '_' || c == '`' || c == '|' ||
+        c == '~' || IsDigit(c) || IsAlpha(c)) {
```
That's a reasonable intuition.  Done.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.h:
```diff
@@ -0,0 +1,293 @@
+//
+//
+// quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
+// qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
```
Done.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+  }
+  if (context.next() == input.next()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected token starting at position %d.", input.next()));
```
Done.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.cc:
```diff
@@ -0,0 +1,357 @@
+
+  if (StatusOr<ParseContext> ipv6 = ParsePlausibleIpV6(context); ipv6) {
+    context.update(*ipv6);
+  } else {
```
Done.

justin-mp commented on test/extensions/filters/http/cdn/status_helpers.h:
```diff
@@ -0,0 +1,34 @@
+#include "gmock/gmock.h"
```
Done.

justin-mp commented on source/extensions/filters/http/cdn/cdn_loop_parser.h:
```diff
@@ -0,0 +1,293 @@
+  constexpr void increment() { ++next_; }
+
+  // Update next from another context.
+  constexpr void update(const ParseContext& other) { next_ = other.next_; }
```
I renamed `update` to `setNext` since that's an even better description of what it does.

penguingao commented on source/extensions/filters/http/cdn_loop/parser.h:
```diff
@@ -0,0 +1,299 @@
+  ParseContext(absl::string_view value, absl::string_view::size_type next)
+      : value_(value), next_(next) {}
+
+  // Return true if we have reached the end of value.
```
nit: here and the rest, s/Return/Returns

https://google.github.io/styleguide/cppguide.html#Function_Comments

justin-mp replied:
Done.

penguingao commented on test/test_common/status_helpers.h:
```diff
@@ -0,0 +1,42 @@
+#pragma once
```
nit: This file is a dup of status_utility.h

justin-mp replied:
It looks like a botched a rename.  Fixed!

penguingao reviewed (approved):
LGTM. A few nits.

alyssawilk commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+        absl::StrFormat("expected cdn-id at position %d; found end-of-input", context.next()));
+  }
+
+  if (StatusOr<ParseContext> ipv6 = parsePlausibleIpV6(context); ipv6) {
```
Are we ever using the status here?

I love the consistency and debugability of StatusOr but I'm less excited about doing multiple string format calls when the headers are valid, and we're trying to sort out which type of valid they are.

Alternately since I think we can determine if it's plausible ipv6 or not from a one character peek, we could do the one character peek and try parsing ipv6 or token based on that?

justin-mp replied:
Using peek worked out nicely.  I did it here and in the values of parameters (where we can parse double-quoted strings).

alyssawilk commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+  } else if (StatusOr<ParseContext> token = parseToken(context); token) {
+    context.setNext(*token);
+  } else {
+    return token.status();
```
Arguably the correct status to return is "couldn't find a token or plausible  ipv6 address" no?

justin-mp replied:
With the new peek code you'll now get something along the lines of "expected IPv6" address after a `[` and "expected token" otherwise, so you get a more specific error rather than just the token's one.

alyssawilk commented on test/extensions/filters/http/cdn_loop/parser_test.cc:
```diff
@@ -0,0 +1,455 @@
+}
+
+TEST(ParseQuotedPairTest, EndOfInput) {
+  const std::string value = R"(a)";
```
What was this one testing? Looks like MissingBackslash

justin-mp replied:
It was supposed to be testing against empty string.  Fixed!

alyssawilk commented on test/extensions/filters/http/cdn_loop/parser_test.cc:
```diff
@@ -0,0 +1,455 @@
+}
+
+TEST(ParseCdnIdTest, SecondInSeries) {
+  // This showcases a bug where I thought absl::string_view::substr took (start,
```
Repharse?
// Make sure that stringpiece code correctly uses (start, end) not (start, len)
?

alyssawilk reviewed (commented):
This is amazing.  Thank you for the incredibly readable code and useful RFC links.

One overall thing to look at is it's a bit below the Envoy coverage bar of 96.6 (Once per-extension coverage checks are fixedby #13030 and this won't merge.)

Mind taking a look at your coverage build results
(https://storage.googleapis.com/envoy-pr/12894/coverage/source/extensions/filters/http/cdn_loop/index.html)
and unit testing the few remaining corner cases and utilities?

justin-mp commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+        !quoted_pair_context) {
+      return quoted_pair_context.status();
+    } else {
+      context.setNext(*quoted_pair_context);
```
All of the parser functions call either `increment` or `setNext` on a context. (In some cases, like parseOptionalWhitespace, a particular invocation may not call either `increment` or `setNext`.)

The problem with passing in the context as a non-const is that undoing the changes is harder since we'd have to wind back any call to `increment` or `setNext`.  Contexts are pretty small (24 bytes on a 64-bit platform), so having a local copy seemed like a good idea to me.

justin-mp commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+//
+// HEXDIG         =  DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
+//
+// This rule allows lower case letters too in violation of the RFC.
```
It's mainly for hex IPv6 addresses.  I updated the comment.

justin-mp commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+         ('\x5D' <= c && c <= '\x7E') || isObsText(c);
+}
+
+// RFC 5324 Appendix B.1 says:
```
Done.

justin-mp commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+//
+// ALPHA          =  %x41-5A / %x61-7A   ; A-Z / a-z
+constexpr bool isAlpha(char c) {
+  return ('\x41' <= c && c <= '\x5a') || ('\x61' <= c && c <= '\x7a');
```
I was trying to be consistent with the RFC grammar, but I could go either way.  (The part after the `;` is a BNF comment, so it's not normative.)

justin-mp commented on source/extensions/filters/http/cdn_loop/parser.cc:
```diff
@@ -0,0 +1,359 @@
+
+} // namespace
+
+ParseContext parseOptionalWhitespace(const ParseContext& input) {
```
`skip` as a prefix is a great suggestion since this is the only "parse" function that can't fail, so it looks different than the other ones.  I went with skipOptionalWhitespace since "optional whitespace" is something the RFC mentions a bunch in contrast to other whitespace.

justin-mp commented:
> One overall thing to look at is it's a bit below the Envoy coverage bar of 96.6 (Once per-extension coverage checks are fixedby #13030 and this won't merge.)

I increased test coverage to 100% according to that link.  (You may have to shift+reload to get the newest figures.)

justin-mp commented:
I think I've addressed all of the comments at this point.  I just merged from master to get the latest changes.  Let me know if there's anything more to do!

lizan commented on CODEOWNERS:
```diff
@@ -23,6 +23,8 @@ extensions/filters/common/original_src @snowp @klarose
 # thrift_proxy extension
 /*/extensions/filters/network/thrift_proxy @zuercher @rgs1
+# cdn_loop extension
+/*/extensions/filters/http/cdn @justin-mp @penguingao @alyssawilk
```
I guess this extension will be in a follow up PR?

justin-mp replied:
Yes and no.

Yes, in that I have a follow-up PR with the cdn_loop filter ready to go.

No, in that the "cdn" above is a typo from before we renamed the filter from `extensions/filters/http/cdn` to `extensions/filters/http/cdn_loop`.  Give me a sec to fix that.

justin-mp replied:
Fixed!

justin-mp commented:
@lizan Friendly ping!  Anything more you need me to do?

alyssawilk reviewed (approved):
Lizan's out and his comments are addressed, so I'm going to go ahead and merge and we can sort out any further comments in a follow-up when Lizan is back.
